app/src/main/java/com/yml/womensafety/AccelerometerService.kt

```diff
@@ -0,0 +1,134 @@
+    private val CHANNEL_ID: String = "ForegroundService"
+    private lateinit var sensorManager: SensorManager
+    private lateinit var accelerometerSensor: Sensor
+    private lateinit var vibrator: Vibrator
+    private var isAccelerometerSensorAvailable: Boolean = false
+    private var itIsNotFirstTime = false
+    private var currentX = 0f
+    private var currentY = 0f
+    private var currentZ = 0f
+    private var lastX = 0f
+    private var lastY = 0f
+    private var lastZ = 0f
+    private val shakeThreshold = 15f
+
+    override fun onCreate() {
+        super.onCreate()
+
+        vibrator = getSystemService(VIBRATOR_SERVICE) as Vibrator
+        sensorManager = getSystemService(SENSOR_SERVICE) as SensorManager
+
+        if (sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) != null) {
+            accelerometerSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
+            isAccelerometerSensorAvailable = true
+        } else {
+            Toast.makeText(
+                applicationContext,
+                "Accelerometer sensor not available",
+                Toast.LENGTH_SHORT
+            ).show()
+            isAccelerometerSensorAvailable = false
+        }
+
+        if (isAccelerometerSensorAvailable) {
+            sensorManager.registerListener(
+                this,
+                accelerometerSensor,
+                SensorManager.SENSOR_DELAY_NORMAL
+            )
+        }
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        createNotificationChannel()
+        val notificationIntent = Intent(this, HomePageActivity::class.java)
+        val pendingIntent: PendingIntent = PendingIntent.getActivity(this, 1, notificationIntent, 0)
+        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
+            .setContentTitle(getString(R.string.app_name))
+            .setContentText(getString(R.string.Shake_device_to_send_location))
+            .setSmallIcon(R.drawable.ic_health_and_safety)
+            .setContentIntent(pendingIntent)
+            .build()
+        startForeground(1, notification)
+        return START_NOT_STICKY
+    }
+
+    override fun onSensorChanged(event: SensorEvent?) {
+        currentX = event!!.values[0]
+        currentY = event.values[1]
+        currentZ = event.values[2]
+
+        if (itIsNotFirstTime) {
+            val xDifference = abs(lastX - currentX)
+            val yDifference = abs(lastY - currentY)
+            val zDifference = abs(lastZ - currentZ)
+
+            if ((xDifference > shakeThreshold && yDifference > shakeThreshold) ||
+                (xDifference > shakeThreshold && zDifference > shakeThreshold) ||
+                (yDifference > shakeThreshold && zDifference > shakeThreshold)
+            ) {
+
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    vibrator.vibrate(
+                        VibrationEffect.createOneShot(
```

Comment:
What happens if OS version is less than O ?

Reply:
It is there only for testing. I will replace that part of code to get location.

Reply:
That should be added as a comment..